src/libmeasurement_kit/ooni/telegram.cpp

```diff
@@ -0,0 +1,194 @@
+// Part of measurement-kit <https://measurement-kit.github.io/>.
+// Measurement-kit is free software. See AUTHORS and LICENSE for more
+// information on the copying conditions.
+
+#include "../common/utils.hpp"
+#include "../ooni/constants.hpp"
+#include <measurement_kit/ooni.hpp>
+
+namespace mk {
+namespace ooni {
+
+using namespace mk::report;
+
+static void tcp_many(const std::vector<std::string> ip_ports,
+                     Var<Entry> entry,
+                     Callback<Error> cb,
+                     Var<Reactor> reactor,
+                     Var<Logger> logger) {
+    // C++20 should have continuations in std; until then,
+    // we copy the ghetto pattern from web_connectivity
+    int ips_count = ip_ports.size();
+    Var<int> ips_tested(new int(0));
+    if (ips_count == *ips_tested) {
+        cb(NoError());
+        return;
+    }
+
+    // per-endpoint, we only consider success/failure.
+    // telegram is "blocked" if no endpoints succeed.
+    auto tcp_cb = [=](std::string ip, int port) {
+        return [=](Error err, Var<net::Transport> txp) {
+            bool close_txp = true; // if connected, we must disconnect
+            Entry result = {
+                {"ip", ip},
+                {"port", port},
+                {"status",
+                    {{"success", nullptr},
+                    {"failure", nullptr}}},
+            };
+            if (!!err) {
+                logger->info("telegram: failure TCP connecting to %s:%d",
+                             ip.c_str(), port);
+                result["status"]["success"] = false;
+                result["status"]["failure"] = err.as_ooni_error();
+                close_txp = false;
+            } else {
+                logger->info("telegram: success TCP connecting to %s:%d",
+                             ip.c_str(), port);
+                result["status"]["success"] = true;
+            }
+            (*entry)["tcp_connect"].push_back(result);
+            *ips_tested += 1;
```

Comment:
>  I guess it's not if we're single-threaded

Yes, there is a single-threaded assumption in many places in the code. Even without multiple threads we may enter into race conditions domain, if we're using Windows with IOCP where basically whenever I/O completes a callback is called (pretty much like signals). The way in which we solve this is by configuring `libevent` to serialize callbacks for us.

In the future, I would like to pay more attention to this potential race conditions, perhaps by creating some abstraction that combines `atomic` and `Var`. Another thing to consider is that, in general, I find that for our use case it would be fine to have a single I/O thread and perhaps a thread pool to hand over heavy work to. Anyway, having the `atomic + Var` thing is perhaps a matter of correctness and I guess it should be done in any case (it will basically boil down to chain an indirect access through `Var` and an atomic access).

> and there are no exceptions

In regards to exceptions, the model that I'm trying to push is the following: in most cases basic sync operations return `ErrorOr<Value>` (if we switch to C++17 this could easily become a tuple), pretty much like `golang` does, because I don't like to *implicit* implied by mixing code that may throw exceptions and code with callbacks. With @AntonioLangiu we were also writing on a more generic template piece of code that basically wraps code that can throw and, depending on usage, either returns or forwards the error that occurred. This way, just by reading the code, one can be sure that exceptions are properly forwarded along the callbacks chain.

> or IO operations in this code?

Speaking of I/O operations, the way in which we're trying to structure them is this:

- if it's a network I/O operation, it has a callback that takes an error as parameter and it should never throw an exception, unless something very bad happens

- otherwise, we generally rely on the fact that C++ streams do not throw by default unless instructed to do so, and then we check the result bits after I/O and act accordingly